File: Tests/MaxEditDraftTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

final class MaxEditDraftTests: XCTestCase {
    func testOpeningAnEditorDoesNotCreateAnotherBenchmark() {
        let draft = MaxEditDraft(leftKg: 31.23456789, rightKg: 34.87654321)
        XCTAssertEqual(draft.leftKg, 31.23456789)
        XCTAssertEqual(draft.rightKg, 34.87654321)
        XCTAssertTrue(draft.changes.isEmpty)
        XCTAssertFalse(draft.canSave)
    }

    func testOnlyTheEditedHandIsSavedIncludingALowerWorkingMax() {
        var draft = MaxEditDraft(leftKg: 40, rightKg: 45)
        draft.leftKg = 35
        XCTAssertEqual(draft.changes, [.init(side: .left, kg: 35)])
        XCTAssertTrue(draft.canSave)
    }

    func testRevertingAnEditDoesNotSaveAnotherRecord() {
        var draft = MaxEditDraft(leftKg: 40, rightKg: 45)
        draft.leftKg = 35
        draft.leftKg = 40
        XCTAssertTrue(draft.changes.isEmpty)
        XCTAssertFalse(draft.canSave)
    }

    func testTwoHandEditsAreIndependentAndNeverCreateASharedMax() {
        var draft = MaxEditDraft()
        draft.leftKg = 24
        draft.rightKg = 28
        XCTAssertEqual(draft.changes, [.init(side: .left, kg: 24), .init(side: .right, kg: 28)])
        XCTAssertFalse(draft.changes.contains { $0.side == .both })
        XCTAssertNil(draft.originalKg(for: .both))
    }

    func testMissingHandIsNotInferredFromTheOtherHand() {
        var draft = MaxEditDraft(leftKg: 40)
        XCTAssertEqual(draft.rightKg, 0)
        XCTAssertNil(draft.originalKg(for: .right))
        draft.leftKg = 42
        XCTAssertEqual(draft.changes, [.init(side: .left, kg: 42)])
        XCTAssertTrue(draft.canSave, "An untouched missing hand must not block the edited hand.")
    }

    func testZeroCannotSilentlyDeleteOneHandWhileSavingTheOther() {
        var draft = MaxEditDraft(leftKg: 40, rightKg: 45)
        draft.leftKg = 0
        draft.rightKg = 47
        XCTAssertTrue(draft.hasInvalidChanges)
        XCTAssertFalse(draft.canSave)
        XCTAssertEqual(draft.changes, [.init(side: .right, kg: 47)])
    }

    func testNonFiniteAndNegativeChangesCannotBeSaved() {
        for invalid in [Double.nan, .infinity, -.infinity, -1] {
            var draft = MaxEditDraft(leftKg: 40)
            draft.leftKg = invalid
            XCTAssertTrue(draft.hasInvalidChanges)
            XCTAssertFalse(draft.canSave)
            XCTAssertTrue(draft.changes.isEmpty)
        }
    }

    func testPoundsDisplayDoesNotChangeStoredPrecision() {
        let kilograms = 31.23456789
        var draft = MaxEditDraft(leftKg: kilograms)
        _ = WeightUnit.lb.fromKg(draft.leftKg)
        XCTAssertTrue(draft.changes.isEmpty)
        XCTAssertEqual(draft.leftKg, kilograms)

        let enteredPounds = 75.5
        draft.rightKg = WeightUnit.lb.toKg(enteredPounds)
        XCTAssertEqual(draft.changes.count, 1)
        XCTAssertEqual(draft.changes.first?.side, .right)
        XCTAssertEqual(draft.changes.first?.kg ?? 0, 34.246223935, accuracy: 1e-12)
        XCTAssertEqual(draft.leftKg, kilograms)
    }

    func testDeletingTheCurrentMaxRefreshesAnUntouchedFieldToItsPredecessor() {
        var draft = MaxEditDraft(leftKg: 40, rightKg: 45)
        draft.rebase(leftKg: 30, rightKg: 45)
        XCTAssertEqual(draft.leftKg, 30)
        XCTAssertEqual(draft.originalKg(for: .left), 30)
        XCTAssertEqual(draft.rightKg, 45)
        XCTAssertTrue(draft.changes.isEmpty)
        XCTAssertFalse(draft.canSave)
    }

    func testDeletingTheCurrentMaxPreservesATypedEditAgainstTheNewPredecessor() {
        var draft = MaxEditDraft(leftKg: 40, rightKg: 45)
        draft.leftKg = 33
        draft.rebase(leftKg: 30, rightKg: 45)
        XCTAssertEqual(draft.leftKg, 33)
        XCTAssertEqual(draft.originalKg(for: .left), 30)
        XCTAssertEqual(draft.rightKg, 45)
        XCTAssertEqual(draft.changes, [.init(side: .left, kg: 33)])
        XCTAssertTrue(draft.canSave)
    }

    func testRebasingAfterTheLastHandRecordIsDeletedDoesNotCopyTheSharedFallback() {
        var current = MaxTable()
        let grip = GripSpec().key
        current.record(50, grip: grip, side: .both)
        current.record(45, grip: grip, side: .right)
        var draft = MaxEditDraft(leftKg: 40, rightKg: 45)
        draft.rebase(leftKg: current.exact(grip: grip, side: .left),
                     rightKg: current.exact(grip: grip, side: .right))
        XCTAssertEqual(current.max(grip: grip, side: .left), 50)
        XCTAssertEqual(draft.leftKg, 0)
        XCTAssertNil(draft.originalKg(for: .left))
        XCTAssertEqual(draft.rightKg, 45)
        XCTAssertTrue(draft.changes.isEmpty)
    }

    func testAConcurrentRecordMatchingTheTypedValueNoLongerNeedsAnotherSave() {
        var draft = MaxEditDraft(leftKg: 40, rightKg: 45)
        draft.leftKg = 33
        draft.rebase(leftKg: 33, rightKg: 45)
        XCTAssertEqual(draft.leftKg, 33)
        XCTAssertEqual(draft.originalKg(for: .left), 33)
        XCTAssertTrue(draft.changes.isEmpty)
        XCTAssertFalse(draft.canSave)
    }
}

File: android/engine/src/test/kotlin/run/nuri/getagrip/engine/MaxEditDraftTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.engine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class MaxEditDraftTests {
    @Test fun openingAnEditorDoesNotCreateRecordsForEitherHand() {
        val draft = MaxEditDraft(30.0, 40.0)
        assertFalse(draft.canSave)
        assertTrue(draft.changes.isEmpty())
        assertEquals(30.0, draft.leftKg)
        assertEquals(40.0, draft.rightKg)
    }

    @Test fun missingHandDoesNotInheritAnotherHandsMax() {
        val draft = MaxEditDraft(leftKg = 30.0)
        assertEquals(0.0, draft.rightKg)
        assertNull(draft.originalKg(Side.right))
        draft.rightKg = 35.0
        assertEquals(listOf(MaxEditDraft.Change(Side.right, 35.0)), draft.changes)
    }

    @Test fun unchangedRetestOnlyIncludesTheExplicitlySelectedHand() {
        val draft = MaxEditDraft(30.0, 40.0)
        draft.setRecordsAnotherTest(true, Side.left)
        assertTrue(draft.canSave)
        assertEquals(listOf(MaxEditDraft.Change(Side.left, 30.0)), draft.changes)
        draft.setRecordsAnotherTest(false, Side.left)
        assertFalse(draft.canSave)
    }

    @Test fun aRetestCannotInventAnUnsetOrSharedHand() {
        val draft = MaxEditDraft(leftKg = 30.0)
        draft.setRecordsAnotherTest(true, Side.right)
        draft.setRecordsAnotherTest(true, Side.both)
        assertTrue(draft.repeatedTests.isEmpty())
        assertFalse(draft.canSave)
    }

    @Test fun anInvalidChangedHandRejectsTheWholeEdit() {
        for (invalid in listOf(0.0, -1.0, Double.NaN, Double.POSITIVE_INFINITY)) {
            val draft = MaxEditDraft(30.0, 40.0)
            draft.leftKg = invalid
            draft.rightKg = 45.0
            assertTrue(draft.hasInvalidChanges)
            assertFalse(draft.canSave)
        }
    }

    @Test fun rebaseFollowsUntouchedValuesAndKeepsExplicitEdits() {
        val draft = MaxEditDraft(30.0, 40.0)
        draft.leftKg = 33.0
        draft.rebase(31.0, 41.0)
        assertEquals(33.0, draft.leftKg)
        assertEquals(41.0, draft.rightKg)
        assertEquals(listOf(MaxEditDraft.Change(Side.left, 33.0)), draft.changes)
    }

    @Test fun rebaseDoesNotReplaceAnExplicitUnchangedRetest() {
        val draft = MaxEditDraft(30.0, 40.0)
        draft.setRecordsAnotherTest(true, Side.left)
        draft.rebase(35.0, 45.0)
        assertEquals(30.0, draft.leftKg)
        assertEquals(45.0, draft.rightKg)
        assertTrue(draft.recordsAnotherTest(Side.left))
    }

    @Test fun aCopyCanBeEditedWithoutMutatingTheDisplayedOriginal() {
        val original = MaxEditDraft(30.0, 40.0)
        val copy = original.copy()
        copy.leftKg = 32.0
        copy.setRecordsAnotherTest(true, Side.right)
        assertEquals(30.0, original.leftKg)
        assertTrue(original.repeatedTests.isEmpty())
        assertFalse(original.canSave)
        assertEquals(2, copy.changes.size)
    }
}
